src/rtc.rs

```diff
@@ -3,131 +3,171 @@
 use crate::datetime::*;
 use crate::rcc::{BDCR, APB1R1, Clocks};
 use crate::pwr;
-use crate::stm32::{RTC};
+use crate::stm32::{RTC, RCC};
 
 /// RTC Abstraction
 pub struct Rtc {
-    rtc: RTC
+    rtc: RTC,
+    rtc_config : RtcConfig
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(u8)]
+pub enum RtcClockSource {
+    /// 00: No clock
+    NoClock = 0b00,
+    /// 01: LSE oscillator clock used as RTC clock
+    LSE = 0b01,
+    /// 10: LSI oscillator clock used as RTC clock
+    LSI = 0b10,
+    /// 11: HSE oscillator clock divided by 32 used as RTC clock
+    HSE = 0b11,
+}
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct RtcConfig {
+    /// RTC clock source
+    clock_config: RtcClockSource,
+    /// Asynchronous prescaler factor
+    /// This is the asynchronous division factor:
+    /// ck_apre frequency = RTCCLK frequency/(PREDIV_A+1)
+    async_prescaler: u8,
+    /// Synchronous prescaler factor
+    /// This is the synchronous division factor:
+    /// ck_spre frequency = ck_apre frequency/(PREDIV_S+1)
+    sync_prescaler: u16,
+}
+
+impl Default for RtcConfig {
+    /// LSI with prescalers assuming 32.768 kHz.
+    /// Raw sub-seconds in 1/256.
+    fn default() -> Self {
+        RtcConfig{
+            clock_config : RtcClockSource::LSI,
+            async_prescaler : 127,
+            sync_prescaler : 255,
+        }
+    }
+}
+
+impl RtcConfig {
+    pub fn clock_config(mut self, cfg : RtcClockSource) -> Self{
+        self.clock_config = cfg;
+        self
+    }
+
+    pub fn async_prescaler(mut self, prescaler : u8) -> Self{
+        self.async_prescaler = prescaler;
+        self
+    }
+
+    pub fn sync_prescaler(mut self, prescaler : u16) -> Self{
+        self.sync_prescaler = prescaler;
+        self
+    }
 }
 
 impl Rtc {
-    pub fn rtc(rtc: RTC, apb1r1: &mut APB1R1, bdcr: &mut BDCR, pwrcr1: &mut pwr::CR1, clocks: Clocks) -> Self {
+    pub fn rtc(rtc: RTC, apb1r1: &mut APB1R1, bdcr: &mut BDCR, pwrcr1: &mut pwr::CR1, clocks: Clocks, rtc_config : RtcConfig) -> Self {
 
-        assert_eq!(clocks.lsi(), true); // make sure LSI is enabled
+        // assert_eq!(clocks.lsi(), true); // make sure LSI is enabled
         // enable peripheral clock for communication
         apb1r1.enr().modify(|_, w| w.rtcapben().set_bit());
         pwrcr1.reg().read(); // read to allow the pwr clock to enable
+
+        let mut rtc_struct = Self {
+            rtc,
+            rtc_config
+        };
+        rtc_struct.set_config(apb1r1, bdcr, pwrcr1, clocks, rtc_config);
```

Comment:
I am not a huge fan of the way this is done. I know it's to allow people to change the config after initialization, but would anyone actually do that in practice? @MabezDev  @korken89 ?

Reply:
In order to react to a CSS event there has to be a way of changing the clock source. This is described as a "must" in the documentation under CSS on LSE in the RCC chapter.